topdown/cache/cache.go

```diff
@@ -165,3 +171,14 @@ func (c *cache) maxSizeBytes() int64 {
 	}
 	return *c.config.InterQueryBuiltinCache.MaxSizeBytes
 }
+
+type keyMutex struct {
+	sync.Map
+}
+
+func (m *keyMutex) Lock(k ast.Value) func() {
+	value, _ := m.LoadOrStore(k, &sync.Mutex{})
+	mtx := value.(*sync.Mutex)
+	mtx.Lock()
+	return func() { mtx.Unlock() }
```

Comment:
[nit] would this do the trick, too?
```suggestion
	return mtx.Unlock
```

Reply:
Done.